src/asm/isr.s

```diff
@@ -5,16 +5,18 @@
 [global interrupt_service_routine_%1]
 interrupt_service_routine_%1:
     cli
+    ; When this macro is called the status registers are already on the stack
     push 0	; since we have no error code, to keep things consistent we push a default EC of 0
     push %1 ; pushing the interrupt number for easier identification by the handler
-	save_context
-	mov rdi, rsp
-    cld
-    call interrupts_handler
-	restore_context
-	add rsp, 16
-	sti
-    iretq
+    save_context ; Now we can save the general purpose registers
+    mov rdi, rsp    ; Let's set the current stack pointer as a parameter of the interrupts_handler
+    cld ; Clear the direction flag
+    call interrupts_handler ; Now we call the interrupt handler
+    mov rsp, rax    ; use the returned context
+    restore_context ; We served the interrupt let's restore the previous context
+    add rsp, 16 ; We can discard the interrupt number and the error code
+    sti
```

Comment:
enabling interrupts inside of an interrupt handler is a bad idea! All your idt entries should be set with interrupt gates, which clears the interrupt flag when the iret frame is pushed onto the stack. iret then pops rflags from the stack, restoring interrupts. So no need to mess with the interrupt flags.